components/script/dom/blob.rs

```diff
@@ -130,212 +64,114 @@ impl Blob {
+            blobs.insert(blob_id.clone(), blob_impl);
+        } else {
+            let mut blobs = HashMap::new();
+            blobs.insert(new_blob_id.clone(), blob_impl);
+            *blob_impls = Some(blobs);
         }
-    }
 
-    /// Cleanups at the time of destruction/closing
-    fn clean_up_file_resource(&self) {
-        if let BlobImpl::File(ref f) = *self.blob_impl.borrow() {
-            let origin = get_blob_origin(&self.global().get_url());
+        let PipelineNamespaceId(name_space) = new_blob_id.namespace_id;
+        let BlobIndex(index) = new_blob_id.index;
+        let index = index.get();
 
-            let (tx, rx) = ipc::channel(self.global().time_profiler_chan().clone()).unwrap();
+        let name_space = name_space.to_ne_bytes();
+        let index = index.to_ne_bytes();
 
-            let msg = FileManagerThreadMsg::DecRef(f.id.clone(), origin, tx);
-            self.send_to_file_manager(msg);
-            let _ = rx.recv();
-        }
+        // 3. Return a u32 pair representation of the key where the object is stored.
+        Ok((u32::from_ne_bytes(name_space), u32::from_ne_bytes(index)))
     }
 
-    fn send_to_file_manager(&self, msg: FileManagerThreadMsg) {
-        let global = self.global();
-        let resource_threads = global.resource_threads();
-        let _ = resource_threads.send(CoreResourceMsg::ToFileManager(msg));
-    }
-}
+    /// <https://html.spec.whatwg.org/multipage/#deserialization-steps>
+    fn deserialize(
+        owner: &DomRoot<GlobalScope>,
+        sc_holder: &mut StructuredDataHolder,
+        extra_data: (u32, u32),
+    ) -> Result<usize, ()> {
+        // 1. Re-build the key for the storage location
+        // of the serialized object.
+        let namespace_id = PipelineNamespaceId(extra_data.0);
+        let index = BlobIndex(NonZeroU32::new(extra_data.1).expect("Index to be non-zero"));
+
+        let id = BlobId {
+            namespace_id,
+            index,
+        };
 
-impl Drop for Blob {
-    fn drop(&mut self) {
-        self.clean_up_file_resource();
-    }
-}
+        let (blobs, blob_impls) = match sc_holder {
+            StructuredDataHolder::Read {
+                blobs, blob_impls, ..
+            } => (blobs, blob_impls),
+            _ => panic!("Unexpected variant of StructuredDataHolder"),
+        };
 
-fn read_file(global: &GlobalScope, id: Uuid) -> Result<Vec<u8>, ()> {
-    let resource_threads = global.resource_threads();
-    let (chan, recv) = ipc::channel(global.time_profiler_chan().clone()).map_err(|_| ())?;
-    let origin = get_blob_origin(&global.get_url());
-    let check_url_validity = false;
-    let msg = FileManagerThreadMsg::ReadFile(chan, id, check_url_validity, origin);
-    let _ = resource_threads.send(CoreResourceMsg::ToFileManager(msg));
+        // 2. Get the transferred object from it's storage, using the key.
+        let (blobs_len, blob_impl) = if let Some(blobs) = blob_impls.as_mut() {
```

Comment:
`blob_impls.as_mut().expect("A blob was deserialized, yet the SC holder does not have any blob impls")`